src/converters/cover.ts

```diff
@@ -252,14 +295,39 @@ class CoverHandler implements ServiceHandler {
       this.current_max
     );
 
-    this.service.updateCharacteristic(hap.Characteristic.CurrentPosition, characteristicValue);
-
-    if (isStopped) {
-      // Update target position and position state
-      // This should improve the UX in the Home.app
-      this.accessory.log.debug(`${this.accessory.displayName}: cover: assume movement stopped`);
+    if (this.motorState === CoverHandler.MOTOR_STATE_CLOSING && this.motorStatePrevious !== CoverHandler.MOTOR_STATE_CLOSING) {
+      this.service.updateCharacteristic(hap.Characteristic.PositionState, hap.Characteristic.PositionState.DECREASING);
+      this.accessory.log.debug(`${this.accessory.displayName}: cover: closing via motor_state`);
+      if (!this.setTargetPositionHandled) {
+        this.service.updateCharacteristic(hap.Characteristic.TargetPosition, 0);
+      }
+      this.motorStatePrevious = CoverHandler.MOTOR_STATE_CLOSING;
+      this.setTargetPositionHandled = false;
+    } else if (this.motorState === CoverHandler.MOTOR_STATE_OPENING && this.motorStatePrevious !== CoverHandler.MOTOR_STATE_OPENING) {
+      this.service.updateCharacteristic(hap.Characteristic.PositionState, hap.Characteristic.PositionState.INCREASING);
+      this.accessory.log.debug(`${this.accessory.displayName}: cover: opening via motor_state`);
+      if (!this.setTargetPositionHandled) {
+        this.service.updateCharacteristic(hap.Characteristic.TargetPosition, 100);
+      }
+      this.motorStatePrevious = CoverHandler.MOTOR_STATE_OPENING;
+      this.setTargetPositionHandled = false;
+    } else if (this.motorState === CoverHandler.MOTOR_STATE_STOPPED && this.motorStatePrevious !== CoverHandler.MOTOR_STATE_STOPPED) {
       this.service.updateCharacteristic(hap.Characteristic.PositionState, hap.Characteristic.PositionState.STOPPED);
+      this.service.updateCharacteristic(hap.Characteristic.CurrentPosition, characteristicValue);
       this.service.updateCharacteristic(hap.Characteristic.TargetPosition, characteristicValue);
+      this.accessory.log.debug(`${this.accessory.displayName}: cover: stopped via motor_state`);
+      this.motorStatePrevious = CoverHandler.MOTOR_STATE_STOPPED;
+      this.setTargetPositionHandled = false;
+    } else if (!this.modern) {
```

Comment:
You can get rid of quite some code duplication here I think.

First of all, I would start by checking we do **NOT** have the `motorStateExpose`. If that's the case we apply the old logic.

Else, we first check if `motorState != motorStatePrevious`. If so:
* `motorStatePrevious = motorState`
* have a temporary value for the new target position value and new PositionState value and use a `switch` on the `motorState` to determine what the values should be for those.
* then update the characteristics accordingly.

Hopefully you get what I mean.